store/notifyer.go

```diff
@@ -0,0 +1,71 @@
+package store
+
+import (
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"sync/atomic"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+)
+
+func (fs *FileStore) startSubDirectoriesFileWatcher() {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	go func() {
+		for {
+
+			// iterate over all subdirectories under root path
+			if err := filepath.Walk(fs.root,
+				func(path string, fi os.FileInfo, err error) error {
+					if fi.Mode().IsDir() {
+						return watcher.Add(path)
+					}
+					return nil
+				}); err != nil {
+
+				log.Fatalf("ERROR: %s", err)
+			}
+
+			// gather events
+			select {
```

Comment:
The code that updates the in-memory metadata looks good, but each time a file is changed, the whole directory tree is walked again. This is very inefficient.

You should walk the entire root once, then walk each new subdirectory *when it's created* to attach listeners as the subdirectories are created.

Reply:
Fixed